crates/ra_hir/src/expr.rs

```diff
@@ -103,6 +107,18 @@ impl BodySyntaxMapping {
     }
 }
 
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Literal {
+    String(String),
+    ByteString(Vec<u8>),
+    Char(char),
+    Bool(bool),
+    Int(u64, UncertainIntTy),
+    Float(u64, UncertainFloatTy), // FIXME: f64 is not Eq
```

Comment:
Do we need to handle `u/i/f128` here?

Reply:
Yeah I think so. Also I wonder how we would handle invalid literals here... maybe it would make more sense to keep the string representation here.

Reply:
I think converting literal text value to the actual literal value should ideally happen at the syntax level (that's what the validators layer is for). We don't know the exact type we need at that time, but I think storing `(bool, u128)` and `f64` should allow us to narrow the type later, including reporting out of range errors?

Reply:
I'm leaving the passing of actual values for another PR since it's not relevant for type inference anyways - currently passing the `Default::default()`.